provider/cmd/shell.go

```diff
@@ -0,0 +1,190 @@
+	"io"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/kubectl/pkg/util/term"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+const (
+	FlagStdin        = "stdin"
+	FlagTty          = "tty"
+	FlagReplicaIndex = "replica-index"
+)
+
+var (
+	errTerminalNotATty = errors.New("Input is not a terminal, cannot setup TTY")
+)
+
+func LeaseShellCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Args:         cobra.MinimumNArgs(2),
+		Use:          "lease-shell",
+		Short:        "do lease shell",
+		SilenceUsage: true,
+		RunE:         doLeaseShell,
+	}
+
+	addLeaseFlags(cmd)
+	cmd.Flags().Bool(FlagStdin, false, "connect stdin")
+	if err := viper.BindPFlag(FlagStdin, cmd.Flags().Lookup(FlagStdin)); err != nil {
+		return nil
+	}
+
+	cmd.Flags().Bool(FlagTty, false, "connect an interactive terminal")
+	if err := viper.BindPFlag(FlagTty, cmd.Flags().Lookup(FlagTty)); err != nil {
+		return nil
+	}
+
+	cmd.Flags().Uint(FlagReplicaIndex, 0, "replica index to connect to")
+	if err := viper.BindPFlag(FlagReplicaIndex, cmd.Flags().Lookup(FlagReplicaIndex)); err != nil {
+		return nil
+	}
+
+	return cmd
+}
+
+func doLeaseShell(cmd *cobra.Command, args []string) error {
+	var stdin io.ReadCloser
+	var stdout io.Writer
+	var stderr io.Writer
+	stdout = os.Stdout
+	stderr = os.Stderr
+	connectStdin := viper.GetBool(FlagStdin)
+	setupTty := viper.GetBool(FlagTty)
+	podIndex := viper.GetUint(FlagReplicaIndex)
+	if connectStdin || setupTty {
+		stdin = os.Stdin
+	}
+
+	var tty term.TTY
+	var tsq remotecommand.TerminalSizeQueue
+	if setupTty {
+		tty = term.TTY{
+			Parent: nil,
+			Out:    os.Stdout,
+			In:     stdin,
+		}
+
+		if !tty.IsTerminalIn() {
+			return errTerminalNotATty
+		}
+
+		dockerStdin, dockerStdout, _ := dockerterm.StdStreams()
+
+		tty.In = dockerStdin
+		tty.Out = dockerStdout
+
+		stdin = dockerStdin
+		stdout = dockerStdout
+		tsq = tty.MonitorSize(tty.GetSize())
+		tty.Raw = true
+	}
+
+	cctx, err := sdkclient.GetClientTxContext(cmd)
+	if err != nil {
+		return err
+	}
+
+	prov, err := providerFromFlags(cmd.Flags())
+	if err != nil {
+		return err
+	}
+
+	bid, err := mcli.BidIDFromFlagsForOwner(cmd.Flags(), cctx.FromAddress)
+	if err != nil {
+		return err
+	}
+
+	cert, err := cutils.LoadAndQueryCertificateForAccount(cmd.Context(), cctx, cctx.Keyring)
+	if err != nil {
+		return err
+	}
+
+	gclient, err := gwrest.NewClient(akashclient.NewQueryClientFromCtx(cctx), prov, []tls.Certificate{cert})
+	if err != nil {
+		return err
+	}
+
+	service := args[0]
+	remoteCmd := args[1:]
+
+	var terminalResizes chan remotecommand.TerminalSize
+	wg := &sync.WaitGroup{}
+	ctx, cancel := context.WithCancel(cmd.Context())
+
+	if tsq != nil {
+		terminalResizes = make(chan remotecommand.TerminalSize, 1)
+		go func() {
+			for {
+				// this blocks waiting for a resize event, the docs suggest
+				// that this isn't the case but there is not a code path that ever does that
+				// so this goroutine is just left running until the process exits
```

Comment:
wat :cry:

Reply:
I read through every code path I could find in the package that defines this. I couldn't find anything that closed the channel, which would unblock the `Next()` method.